File: January2022/C++/String_To_Integer.cpp
class Solution {
public:
    int myAtoi(string s) {
        // return 0;
        long double res=0;
        bool sign=false,setres=false;
        int val=1,i=0;
        
        while(i<s.length() && isspace(s[i])) i++;
        if(!(isdigit(s[i]) ||( (s[i]=='+' || s[i]=='-' )&& isdigit(s[i+1])))) return 0;
        
        while(i<s.length() && (isdigit(s[i]) || !setres))
        {
            if((s[i]=='-' || s[i]=='+') && !sign)
            {
                sign =true;
                val=(s[i]=='-')?-1 : 1;
            }
            if(isdigit(s[i]))
            {
                setres=true;
                res=res*10+(s[i]-'0');
            }
            i++;
        }
        res= val * res;
        return res<INT_MIN?INT_MIN:res>INT_MAX?INT_MAX:(int)res;
    }
};

File: January2022/Python/StringToInteger.py
# Author :  Karthik K (https://www.linkedin.com/in/karthikmunirathinam/)

# TC : O(N)

# SC : O(1)

class Solution:
    def myAtoi(self, s: str) -> int:
        sign = 1
        num = 0
        ind = 0
        INT_MAX = 2**31-1
        INT_MIN = -(2**31)
        while ind < len(s) and s[ind] == ' ':
            ind += 1
        if ind >= len(s):
            return 0
        if s[ind] == '+' or s[ind] == '-':
            sign = 1 if s[ind] == '+' else -1
            ind += 1
        if ind >= len(s):
            return 0
        while ind < len(s) and s[ind].isdigit():
            if num > INT_MAX//10 or (num == INT_MAX//10 and int(s[ind]) > INT_MAX % 10):
                return INT_MAX if sign == 1 else INT_MIN
            num = num*10+int(s[ind])
            ind += 1
        return num*sign
